### ISSUE TITLE
Label colour does not always match colour swatch in the layer controls

### ISSUE THREAD

**Issue by GenevieveBuckley** [napari/napari#5507]
## 🐛 Bug

When shuffling the label colours, every so often the label colour swatch shown in the layer controls on the left hand side does not match the colour of the label(s) in the viewer canvas.

## To Reproduce

The shuffle colour behaviour ~~does not seem to be~~ is not deterministic (EDIT: it's using the `numpy.random.rand()` function [here](https://github.com/napari/napari/blob/180992f2dde1d70e0a89628216b4ca6d56943e5c/napari/layers/labels/labels.py#L965-L966)). This is frustrating, but here's how I'm able to (eventually) see the label colour mismatch.

First, set up a napari viewer:
```python
import numpy as np
import napari

arr = np.random.random((50,50,50))
viewer = napari.view_image(arr)
viewer.add_labels(np.zeros(arr.shape).astype(int))
```

Second, click the canvas to draw a label on it.

Third, click the "shuffle" button repeatedly to shuffle the label display colours. Eventually, a shuffle will result in a label colour swatch (in the controls next to the label id number) that doesn't match the actual colour of the label displayed in the viewer canvas.

EDIT: I've found some "bad seed" values that reliably produce the problem

```python
layer = viewer.layers[-1]

layer.seed = 0.8919032207952365  # brown vs blue

layer.seed = 0.07731549163628793  # red vs brown

layer.seed = 0.09858048620725668  # blue vs brown

layer.seed = 0.4595883721861045  # yellow-ish vs lilac
```

Here are some examples of this behaviour in the wild:

Olive vs brown
![screenshot-napari-label-color-mismatch-shuffle9](https://user-images.githubusercontent.com/30920819/214469942-45b47c3e-ff9d-452c-8074-e8355b04f70e.png)
Green vs tan
![screenshot-napari-label-color-mismatch-shuffle10](https://user-images.githubusercontent.com/30920819/214469946-f6e2f5d8-2343-4e0f-81f3-e2026d2d1381.png)
Purple vs lime green
![screenshot-napari-label-color-mismatch-shuffle19](https://user-images.githubusercontent.com/30920819/214470062-0552288e-30a2-451e-8bf4-2f756e18a038.png)

## Expected behavior

I expect that the colour swatch should always match the colour of the label in the image canvas.

## Environment

napari: 0.4.13.dev917+g54c1dbf8
Platform: macOS-12.3.1-arm64-arm-64bit
System: MacOS 12.3.1
Python: 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:25:29) [Clang 14.0.6 ]
Qt: 5.15.6
PyQt5: 5.15.7
NumPy: 1.24.1
SciPy: 1.10.0
Dask: 2023.1.0
VisPy: 0.12.1
magicgui: 0.6.1
superqt: unknown
in-n-out: 0.1.6
app-model: 0.1.1
npe2: 0.6.2

OpenGL:
- GL version: 2.1 Metal - 76.3
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1440x900, scale 2.0
- screen 2: resolution 1920x1080, scale 1.0

Settings path:
- /Users/genevieb/Library/Application Support/napari/napari-dev_8a798ce8db3962d716d33109f1c23fa2f783f2dd/settings.yaml

## Additional context

Some relevant bits of the code:
* [shuffle colormap button](https://github.com/napari/napari/blob/180992f2dde1d70e0a89628216b4ca6d56943e5c/napari/_qt/layer_controls/qt_labels_controls.py#L163-L170)
* [Qt changeColor](https://github.com/napari/napari/blob/180992f2dde1d70e0a89628216b4ca6d56943e5c/napari/_qt/layer_controls/qt_labels_controls.py#L327-L330)
* [The new_colormap() function that changes the layer seed value](https://github.com/napari/napari/blob/180992f2dde1d70e0a89628216b4ca6d56943e5c/napari/layers/labels/labels.py#L965-L966)

**Comment by jni** [comment]
Damn — this seems to be a duplicate of #3174 which was supposedly fixed in #4321. But I can confirm the behaviour on main locally. Here is a faster reproducer:

```python
import numpy as np
import napari

viewer = napari.view_labels(
        np.ones((5, 5), dtype=int),
        seed=0.8919032207952365,
        )

napari.run()
```

<img width="1281" alt="Screen Shot 2023-01-25 at 9 07 39 pm" src="https://user-images.githubusercontent.com/492549/214510907-4b9eb2de-ba3a-46aa-80d2-0401d0b87a95.png">

**Comment by Czaki** [comment]
It is a known bug. I will try to find the proper issue (if exist). CC @perlman

**Comment by jni** [comment]
It's also worth checking out the branch in #3308 to see if the problem persists there. That branch still has some work left to do but it probably represents the future of the color rendering for labels more closely than main.

**Comment by GenevieveBuckley** [comment]
Maybe a crazy hypothesis, but the `layer.seed` value is float64. Could the colour swatch somehow be using a different datatype without us noticing, like float32 or something?

**Comment by dalthviz** [comment]
Hi, not totally sure if is the right approach but changing the `label_colormap` function to use as colormap interpolation `linear` seems like fixes this? So changing the function at: 

https://github.com/napari/napari/blob/781ec6af2b1fa686afb26478c6461d87aba3624f/napari/utils/colormaps/colormap_utils.py#L399-L440

To be:


```python
def label_colormap(num_colors=256, seed=0.5):
    """Produce a colormap suitable for use with a given label set.

    Parameters
    ----------
    num_colors : int, optional
        Number of unique colors to use. Default used if not given.
        Colors are in addition to a transparent color 0.
    seed : float or array of float, length 3
        The seed for the random color generator.

    Returns
    -------
    colormap : napari.utils.Colormap
        A colormap for use with labels remapped to [0, 1].

    Notes
    -----
    0 always maps to fully transparent.
    """
    # Starting the control points slightly above 0 and below 1 is necessary
    # to ensure that the background pixel 0 is transparent
    midpoints = np.linspace(0.00001, 1 - 0.00001, num_colors)
    control_points = np.concatenate(([0], midpoints, [1.0]))
    # make sure to add an alpha channel to the colors
    colors = np.concatenate(
        (
            _color_random(control_points.size, seed=seed),
            np.full((control_points.size, 1), 1),
        ),
        axis=1,
    )
    # Insert alpha at layer 0
    colors[0, :] = 0  # ensure alpha is 0 for label 0

    return Colormap(
        name='label_colormap',
        display_name=trans._p('colormap', 'low discrepancy colors'),
        colors=colors,
        controls=control_points,
        interpolation='linear',
    )
```

With that, the example code above gives me this:

![imagen](https://user-images.githubusercontent.com/16781833/235232934-39b498dd-ada0-47ff-9450-b4ba60344742.png)

**Comment by psobolewskiPhD** [comment]
I can confirm that the fix by @dalthviz resolves the seed issues from OP and the snippet by @jni (which I can reproduce on main) 
I don't notice any regression in behavior and relevant tests pass.

cc: @perlman

**Comment by GenevieveBuckley** [comment]
Sounds good!
@dalthviz would you like to make a PR, or are you too busy and would prefer for someone else to do that?

**Comment by jni** [comment]
I'm worried that this fix is not quite correct. Linear interpolation is absolutely not what you want in labels colormaps (which are completely discontinuous). So I would prefer we took time to understand the failure and potential fix rather than "🤷‍♂️ it works I think!" (iirc #3308 does fix this btw.)

**Comment by psobolewskiPhD** [comment]
I just tested on main 737ae894b3e80d3bfdef47fd216811c7795042a6 and 0.4.19rc6
(so post #3308 and all the other recent label changes).
I tried all the "bad seeds" from above and could not reproduce e.g. 0.8919032207952365 (@jni script above)

<img width="1198" alt="image" src="https://github.com/napari/napari/assets/76622105/4bf71356-61fa-48de-a39d-4c00ffc40156">

I spammed shuffle too. Going to close!
